evm_arithmetization/src/cpu/kernel/interpreter.rs

```diff
@@ -164,72 +139,54 @@ pub(crate) fn simulate_cpu_and_get_user_jumps<F: Field>(
     }
 }
 
-/// Different types of Memory operations in the interpreter, and the data
-/// required to revert them.
-enum InterpreterMemOpKind {
-    /// We need to provide the context.
-    Push(usize),
-    /// If we pop a certain value, we need to push it back to the correct
-    /// context when reverting.
-    Pop(U256, usize),
-    /// If we write a value at a certain address, we need to write the old value
-    /// back when reverting.
-    Write(U256, usize, usize, usize),
+/// Different types of Memory operations in the interpreter.
+#[derive(Debug)]
+pub(crate) enum InterpreterMemOpKind {
+    Read(U256, MemoryAddress),
+    Write(U256, MemoryAddress),
 }
 
-impl<'a, F: Field> Interpreter<'a, F> {
-    pub(crate) fn new_with_kernel(initial_offset: usize, initial_stack: Vec<U256>) -> Self {
-        let mut result = Self::new(
-            &KERNEL.code,
-            initial_offset,
-            initial_stack,
-            &KERNEL.prover_inputs,
-        );
-        result.initialize_rlp_segment();
-        result
-    }
-
+impl<F: Field> Interpreter<F> {
     /// Returns an instance of `Interpreter` given `GenerationInputs`, and
     /// assuming we are initializing with the `KERNEL` code.
     pub(crate) fn new_with_generation_inputs_and_kernel(
         initial_offset: usize,
         initial_stack: Vec<U256>,
         inputs: GenerationInputs,
     ) -> Self {
-        let mut result = Self::new_with_kernel(initial_offset, initial_stack);
+        let mut result = Self::new(initial_offset, initial_stack);
         result.initialize_interpreter_state_with_kernel(inputs);
         result
     }
 
-    pub(crate) fn new(
-        code: &'a [u8],
-        initial_offset: usize,
-        initial_stack: Vec<U256>,
-        prover_inputs: &'a HashMap<usize, ProverInputFn>,
-    ) -> Self {
+    pub(crate) fn new(initial_offset: usize, initial_stack: Vec<U256>) -> Self {
         let mut result = Self {
-            generation_state: GenerationState::new(GenerationInputs::default(), code)
+            generation_state: GenerationState::new(GenerationInputs::default(), &KERNEL.code)
                 .expect("Default inputs are known-good"),
-            prover_inputs_map: prover_inputs,
             // `DEFAULT_HALT_OFFSET` is used as a halting point for the interpreter,
             // while the label `halt` is the halting label in the kernel.
             halt_offsets: vec![DEFAULT_HALT_OFFSET, KERNEL.global_labels["halt"]],
             halt_context: None,
             debug_offsets: vec![],
-            running: false,
             opcode_count: [0; 256],
             memops: vec![],
             jumpdest_table: HashMap::new(),
+            preinitialized_segments: HashMap::default(),
+            is_jumpdest_analysis: false,
+            clock: 0,
```

Comment:
I wonder if you can't use the default struct initialization syntax here (and elsewhere?): `let a = MyStruct { some_member: 10, another: 'x', ..Default::default() };`?

Reply:
I like it for instance in tests when we often instantiate `BlockMetadata` or similar structs with unimportant fields, but here it may be helpful to clearly see how each internal field is being pre-initialized?